### PR TITLE
Fix/asg resource tags

### LINKED ISSUE #5164: Extended resources provided by ASG via tags is not working
**Which component are you using?**:
autoscaler

**What version of the component are you using?**:
 1.25.0-alpha.0 AND 1.23.1

**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.9", GitCommit:"6df4433e288edc9c40c2e344eb336f63fad45cd2", GitTreeState:"clean", BuildDate:"2022-04-13T19:57:43Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.9", GitCommit:"c1de2d70269039fe55efb98e737d9a29f9155246", GitTreeState:"clean", BuildDate:"2022-07-13T14:19:57Z", GoVersion:"go1.17.11", Compiler:"gc", Platform:"linux/amd64"}

</pre></details>

**What environment is this in?**:
sanbox ( AWS )

**What did you expect to happen?**:
I tried to use extended resource defined as tag in ASG according documentation for AWS it should be `k8s.io/cluster-autoscaler/node-template/resources/<resource-name>` https://github.com/kubernetes/autoscaler/blob/cluster-autoscaler-1.23.1/cluster-autoscaler/cloudprovider/aws/README.md#auto-discovery-setup . This sould work at least for node-group-auto-discovery mode. Is anybody successfully using it?

**What happened instead?**:
This tag is never read or ignored. CA is still complaining about insufficient resource and not scaling up.

**How to reproduce it (as minimally and precisely as possible)**:
Add extended resource as descirbe in https://kubernetes.io/docs/tasks/configure-pod-container/extended-resource/
Add  `k8s.io/cluster-autoscaler/node-template/resources/<resource-name>` tag with same reasonable value to ASG

**Anything else we need to know?**:
I did some tests, it looks like this is never executed https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L412

### PR DESCRIPTION
**Pull request by tombokombo**
#### Which component this PR applies to?
autoscaler

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Problem is described in https://github.com/kubernetes/autoscaler/issues/5164
TLDR 
In [autoscaler docs](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md#auto-discovery-setup) is written
```From version 1.14, Cluster Autoscaler can also determine the resources provided by each Auto Scaling Group via tags. The tag is of the format k8s.io/cluster-autoscaler/node-template/resources/<resource-name>. <resource-name> is the name of the resource, such as ephemeral-storage. The value of each tag specifies the amount of resource provided. The units are identical to the units used in the resources field of a Pod specification.```

But I found no way how to make this work. There is long time relevant part of code but never executed. I'm using kops AWS cluster. So this PR reintroduce this functionality.

#### Which issue(s) this PR fixes:

Fixes #5164

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by tombokombo** [comment]
@gjtempleton are you planning to review this PR?

**Comment by gjtempleton** [comment]
Hey @tombokombo, apologies, I largely use assignment to PRs as my filter for which PRs to review hence me missing it in the GH notifications noise before.

/assign @gjtempleton

**Comment by tombokombo** [comment]
@gjtempleton still nothing?

**Comment by gjtempleton** [comment]
Hey, I've tried to spend some time reproducing the issue this is meant to resolve this evening and haven't been able to reproduce this as long as I updated the existing nodes in the ASG to present the custom resource. Are you able to provide further instructions for reproducing the issue you're seeing?

For reference, I was using a custom resource under the name `example.com/customresource`.

**Comment by paleozogt** [comment]
@gjtempleton If you have nodes running then the problem won't happen.  The issue is what autoscaler does when the ASG is at zero and has never been scaled up before.

**Comment by gjtempleton** [comment]
OK, understood and now reproduced.

I'd like to get @x13n's take here, as if we merged this, we'd be at the point of having two cloud provider specific processors in core code, which I'm pretty wary about. I'd also like to ensure we have some tests for this functionality as we already have around the processing of other resources.

I'm wondering whether this also might be affecting any other cloud providers supporting specifying information like this for scale from zero?

**Comment by paleozogt** [comment]
@gjtempleton I can't comment on this MR, but I'd like to note that this problem has been reported over and over (e.g., #3802, #5006, #5164, #5278, etc).  Its like the feature just wasn't implemented?

**Comment by tombokombo** [comment]
@gjtempleton I've added some tests, but functionality was already there and tested with ephemeral-storage. Regarding implementation, auto-scaling group tags are aws specific, there is autoscaler aws specific documentation that custom-resources should work via tags. That is the reason why I made aws specific provider modifications. I've could put it into generic template-info-provider, but it could brake half of other cloudproviders that I'm not able to test. There is already specific processor for gcp, so I'm not breaking any pattern. This patch could fix aws specific issues mention by @paleozogt

**Comment by x13n** [comment]
I think the way this is introduced into core is fine - we already are doing roughly the same thing with other cloud providers. I don't like this, but wouldn't block this PR on following an already established pattern. That being said, we should really look into better isolation between different cloud providers. https://github.com/kubernetes/autoscaler/issues/5394 I proposed some time ago was - in retrospect - too radical, but I think separating each cloud provider to have its own container image (and hence - own main.go) is the direction into which I'd go. I'm planning to update that issue with more details when I find a little bit of extra time.

So - no blockers from me, but leaving the lgtm and approval to people more familiar with aws.

**Comment by tombokombo** [comment]
@x13n thanks for review. I thought the same when I first dig into the code, that it needs core library and separate cloudproviders code...anyway, @gjtempleton  its your turn now.

**Comment by gjtempleton** [comment]
If you're happy with it as is (and addressing for the long term in the linked issue) then I'm happy to go with this approach.

/approve

**Comment by tombokombo** [comment]
@gjtempleton ok, thx and who can put here lgtm?

**Comment by gjtempleton** [comment]
Ah, sorry, my bad...
/lgtm

**Comment by tombokombo** [comment]
@jbartosik i'm going to provide fix.

**Comment by tombokombo** [comment]
@jbartosik fix here https://github.com/kubernetes/autoscaler/pull/5485
